google/cloud/storage/internal/async/insert_object.cc

```diff
@@ -0,0 +1,121 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/async/insert_object.h"
+#include "google/cloud/storage/internal/crc32c.h"
+#include "google/cloud/storage/internal/grpc/object_request_parser.h"
+#include <type_traits>
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+template <typename T>
+struct AsContentType;
+
+template <>
+struct AsContentType<std::string> {
+  // NOLINTNEXTLINE(performance-unnecessary-value-param)
+  std::string operator()(absl::Cord c) { return static_cast<std::string>(c); }
+};
+
+template <>
+struct AsContentType<absl::Cord> {
+  absl::Cord operator()(absl::Cord c) { return c; }
```

Comment:
`std::move(c)`?

Reply:
I think that would disable copy elision.  The returned value must be a name.